assignment2/2b/src/Tree.java

```diff
@@ -0,0 +1,122 @@
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Tree for getting common ancestors
+ */
+public class Tree<T> {
+    private T value;
+    private Tree<T> parent;
+    private Tree<T> left;
+    private Tree<T> right;
+
+    Tree(T value) {
+        this(value, null, null, null);
+    }
+
+    Tree(T value, Tree<T> parent, Tree<T> left, Tree<T> right) {
+        this.value = value;
+        this.parent = parent;
+        this.left = left;
+        this.right = right;
+    }
+
+    /**
+     * Add new element
+     *
+     * @param newValue  -- new element
+     * @param parentVal -- value of parent
+     * @param isLeft    -- is left or right child
+     * @return true if added
+     */
+    boolean add(T newValue, T parentVal, boolean isLeft) {
+        if (parentVal.equals(value)) {
+            if (isLeft) {
+                left = new Tree<T>(newValue, this, left, null);
+                if (left.left != null) {
+                    left.left.parent = left;
+                }
+            } else {
+                right = new Tree<T>(newValue, this, null, right);
+                if (right.right != null) {
+                    right.right.parent = right;
+                }
+            }
+            return true;
+        } else {
+            return (left != null && left.add(newValue, parentVal, isLeft))
```

Comment:
I know short-circuiting will make it do the right thing here, but stylistically I don't like the idea that we're using what looks like a condition-testing expression to do side-effecting modification. I recommend breaking it out into something like "boolean added = false; if (left != null) { added = left.add(...); } if (right != null) { added |= right.add(...); } return added;" It's pretty ugly, but a little clearer in what it's doing.